streaming/src/main/scala/org/apache/spark/streaming/receiver/ReceiverSupervisor.scala

```diff
@@ -117,11 +117,15 @@ private[streaming] abstract class ReceiverSupervisor(
   /** Start receiver */
   def startReceiver(): Unit = synchronized {
     try {
-      logInfo("Starting receiver")
-      receiver.onStart()
-      logInfo("Called receiver onStart")
-      onReceiverStart()
-      receiverState = Started
+      if (onReceiverStart()) {
+        logInfo("Starting receiver")
+        receiverState = Started
+        receiver.onStart()
+        logInfo("Called receiver onStart")
+      } else {
+        // The driver refused us
+        stop("Registered unsuccessfully because the driver refused" + streamId, None)
+      }
     } catch {
       case t: Throwable =>
```

Comment:
+1. I learnt that the hard way too. See https://www.sumologic.com/2014/05/05/why-you-should-never-catch-throwable-in-scala/ and http://www.scala-lang.org/api/current/scala/util/control/ControlThrowable.html